File: crates/osrep-cli/src/help.rs
//! `--version` and `--help`.
//!
//! The text is the C++'s (`dup_wrapper.cpp:412-472`, which is what a user sees:
//! the wrapper answers both flags before `srep_main` ever runs). A drop-in
//! replacement that prints a different synopsis is a worse drop-in, so this is
//! reproduced rather than rewritten.

use crate::VERSION;

pub const DESCRIPTION: &str =
    "huge-dictionary LZ77 preprocessor (Omega lineage, fork of SREP by Bulat Ziganshin)";
pub const HOMEPAGE: &str = "https://github.com/YadeWira/omega-srep";
/// Release date. No longer `srep.cpp:6`'s `program_date`: the C++ is frozen at
/// 1.0.7 and its date with it (see the note in the workspace `Cargo.toml`).
pub const DATE: &str = "September 22, 2026";

/// `print_version` (`dup_wrapper.cpp:406-409`).
pub fn version() -> String {
    format!("Omega SREP {VERSION} ({DATE})")
}

/// `print_help` (`dup_wrapper.cpp:412-472`).
pub fn help() -> String {
    format!(
        "{version}\n\
         {DESCRIPTION}\n\
         {HOMEPAGE}\n\
         \n\
         Usage:\n\
         \x20 osrep [options] -mN  input  output       compress\n\
         \x20 osrep [options] -d   archive output      decompress\n\
         \x20 osrep -i archive                         show archive info\n\
         \n\
         Compression methods:\n\
         \x20 -m0       REP only (in-memory; works with -dBYTES)\n\
         \x20 -m1, -m2  Content-defined chunking (CDC)\n\
         \x20 -m3       Digest-comparison match search (default)\n\
         \x20 -m4       Full match search (best ratio for general data)\n\
         \x20 -m5       Exhaustive match search (slowest, best ratio)\n\
         \n\
         Common options:\n\
         \x20 -lN, -cN          minimum-match / chunk-size tuning\n\
         \x20 -bN               buffer size (default 8mb)\n\
         \x20 -tN               worker thread count (-m1/-m2's CDC, and -m3/-m5's\n\
         \x20                    prepare_buffer stripe pool, capped at 16 internally)\n\
         \x20 -dBYTES           dictionary size for -m0\n\
         \x20 -hash=NAME        select hash (vmac, sha1, ...)\n\
         \x20 -mmap, -nommap    enable/disable POSIX mmap reads\n\
         \x20 -temp=PATH        tempfile path override\n\
         \x20 --format=v4|v5    container to write (default v5). v4 is the\n\
         \x20                   container the 1.0.x releases read, kept for\n\
         \x20                   interoperability; both are always readable.\n\
         \x20 --seed=N          deterministic hash seed (uint64; default is\n\
         \x20                   random per-run). Same seed + same input ->\n\
         \x20                   byte-identical archive. Useful for content-\n\
         \x20                   addressed storage and reproducible builds.\n\
         \x20 -bar              emit \"PROGRESS <done> <total>\" to stderr every\n\
         \x20                   ~0.5s, plus a guaranteed final line at\n\
         \x20                   done==total. Machine-parseable, for wrapping\n\
         \x20                   osrep as a subprocess.\n\
         \n\
         Dedup pre-pass (Omega F5):\n\
         \x20 -dup              enable FA-style dedup pre-pass\n\
         \x20 --dup-paranoid    byte-compare on every dedup hit (slower, no\n\
         \x20                   silent corruption on 64-bit hash collision)\n\
         \x20 --chunk-avg=N     CDC average chunk size (default 4096)\n\
         \x20 --chunk-min=N     CDC minimum chunk size (default 1024)\n\
         \x20 --chunk-max=N     CDC maximum chunk size (default 16384)\n\
         \x20 --chunk-buf=N     buffer-bounded CDC, bytes (default 8388608)\n\
         \x20 --chunk-hash=NAME CDC boundary hash: fnv (default) or gear.\n\
         \x20                   gear = Gear-hash CDC with an implicit ~64-byte\n\
         \x20                   content window (FastCDC-style normalized\n\
         \x20                   chunking); finds non-buffer-aligned duplicates\n\
         \x20                   that fnv's unwindowed rolling hash misses.\n\
         \x20                   Opt-in; does not change the .dupref format.\n\
         \n\
         Help:\n\
         \x20 --help, -h, -?    this synopsis\n\
         \x20 --version, -V     version line\n\
         \n\
         Project page: {HOMEPAGE}\n",
        version = version(),
    )
}

File: pascal/src/help.pas
unit Help;
{ `--version` and `--help`, byte-for-byte what the C++ prints and the Rust port
  reproduces (`crates/osrep-cli/src/help.rs`, `dup_wrapper.cpp:412-472`).

  A drop-in replacement that prints a different synopsis is a worse drop-in, so
  this is reproduced rather than rewritten. `tests/pascal_cli_conformance.sh`
  diffs it against the shipped Rust binary and fails on a single byte. }

{$MODE OBJFPC}{$H+}
interface

const
  { Kept in step with the released version. The Pascal port reports the same
    line as the Rust one while it is being validated against it; once it is
    what ships, this is what a release bumps. }
  VERSION     = '2.1.0';
  RELEASE_DATE = 'September 22, 2026';
  DESCRIPTION = 'huge-dictionary LZ77 preprocessor (Omega lineage, fork of SREP by Bulat Ziganshin)';
  HOMEPAGE    = 'https://github.com/YadeWira/omega-srep';

function VersionLine: AnsiString;
function HelpText: AnsiString;

implementation

function VersionLine: AnsiString;
begin
  Result := 'Omega SREP ' + VERSION + ' (' + RELEASE_DATE + ')';
end;

{ The synopsis, line by line. LF only -- see Out.WriteRaw for why this never
  goes through a text file. }
function HelpText: AnsiString;
const
  LF = #10;
begin
  Result :=
    VersionLine + LF +
    DESCRIPTION + LF +
    HOMEPAGE + LF +
    LF +
    'Usage:' + LF +
    '  osrep [options] -mN  input  output       compress' + LF +
    '  osrep [options] -d   archive output      decompress' + LF +
    '  osrep -i archive                         show archive info' + LF +
    '  osrep --verify archive                   check it without unpacking' + LF +
    LF +
    'Compression methods:' + LF +
    '  -m0       REP only (in-memory; works with -dBYTES)' + LF +
    '  -m1, -m2  Content-defined chunking (CDC)' + LF +
    '  -m3       Digest-comparison match search (default)' + LF +
    '  -m4       Full match search (best ratio for general data)' + LF +
    '  -m5       Exhaustive match search (slowest, best ratio)' + LF +
    LF +
    'Common options:' + LF +
    '  -lN, -cN          minimum-match / chunk-size tuning' + LF +
    '  -bN               buffer size (default 8mb)' + LF +
    '  -dBYTES           dictionary size for -m0' + LF +
    '  -hash=NAME        select hash (vmac, sha1, ...)' + LF +
    '  -mmap, -nommap    enable/disable POSIX mmap reads' + LF +
    '  -temp=PATH        tempfile path override' + LF +
    '  --format=v4|v5    container to write (default v5). v4 is the' + LF +
    '                    container the 1.0.x releases read, kept for' + LF +
    '                    interoperability; both are always readable.' + LF +
    '  --seed=N          deterministic hash seed (uint64; default is' + LF +
    '                    random per-run). Same seed + same input ->' + LF +
    '                    byte-identical archive. Useful for content-' + LF +
    '                    addressed storage and reproducible builds.' + LF +
    '  -bar              emit "PROGRESS <done> <total>" to stderr every' + LF +
    '                    ~0.5s, plus a guaranteed final line at' + LF +
    '                    done==total. Machine-parseable, for wrapping' + LF +
    '                    osrep as a subprocess.' + LF +
    LF +
    'Dedup pre-pass (Omega F5):' + LF +
    '  -dup              enable FA-style dedup pre-pass' + LF +
    '  --dup-paranoid    byte-compare on every dedup hit (slower, no' + LF +
    '                    silent corruption on 64-bit hash collision)' + LF +
    '  --chunk-avg=N     CDC average chunk size (default 4096)' + LF +
    '  --chunk-min=N     CDC minimum chunk size (default 1024)' + LF +
    '  --chunk-max=N     CDC maximum chunk size (default 16384)' + LF +
    '  --chunk-buf=N     buffer-bounded CDC, bytes (default 8388608)' + LF +
    '  --chunk-hash=NAME CDC boundary hash: fnv (default) or gear.' + LF +
    '                    gear = Gear-hash CDC with an implicit ~64-byte' + LF +
    '                    content window (FastCDC-style normalized' + LF +
    '                    chunking); finds non-buffer-aligned duplicates' + LF +
    '                    that fnv''s unwindowed rolling hash misses.' + LF +
    '                    Opt-in; does not change the .dupref format.' + LF +
    LF +
    'Accepted and not acted on:' + LF +
    '  -tN               the implementation is single-threaded per block,' + LF +
    '                    so there is no worker count to set' + LF +
    '  -aN[/M]           accelerator sizing; allocates nothing here' + LF +
    '  -mmap, -nommap    mmap reads' + LF +
    '  -ia-, -ia+        I/O acceleration' + LF +
    '  -slp, -slp-       large-page mode' + LF +
    '  -pc[N]            progress counters' + LF +
    '  These are parsed and validated so scripts written against the' + LF +
    '  1.0.x C++ keep working unchanged. None of them changes the' + LF +
    '  archive: that was measured before the port, not assumed.' + LF +
    LF +
    'Checking:' + LF +
    '  --verify          v5 only: check an archive''s framing, CRCs and' + LF +
    '                    records without decompressing it. Catches' + LF +
    '                    truncation and framing damage in the time it' + LF +
    '                    takes to read the file. It does NOT cover the' + LF +
    '                    stored block bytes -- damage inside a literal' + LF +
    '                    run still needs a -d. v1-v4 carry no checksum' + LF +
    '                    at all and are refused.' + LF +
    LF +
    'Help:' + LF +
    '  --help, -h, -?    this synopsis' + LF +
    '  --version, -V     version line' + LF +
    LF +
    'Project page: https://github.com/YadeWira/omega-srep' + LF;
end;

end.
